<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto</font></h1>
<h2 align=center><font size = 3>Notebook focusing on segmenting and clustering</font></h2>


## Overview


This is the third notebook of the assignment for "Segmenting and Clustering Neighborhoods in Toronto" using the data obtained from the privious exercises.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item1">Download and Explore Dataset</a>

2. <a href="#item2">Explore Neighborhoods in Toronto</a>

3. <a href="#item3">Analyze Each Neighborhood</a>

4. <a href="#item4">Cluster Neighborhoods</a>

5. <a href="#item5">Examine Clusters</a>    
</font>
</div>

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Solving environment: ...working... done

# All requested packages already installed.

Solving environment: ...working... done

# All requested packages already installed.

Libraries imported.


<a id='item1'></a>

## 1. Download and Explore Dataset

The following exercise uses the data file generated from the previous steps.


In [2]:
# read the csv file 
toronto_data = pd.read_csv("toronto_loc_data.csv")

In [4]:
print("dimension of data:", toronto_data.shape)
toronto_data.head()

dimension of data: (103, 5)


PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

#### Use geopy library to get the latitude and longitude values of city of Toronto.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [5]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Create a map of Toronto  with neighborhoods superimposed on top.

In [6]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [16]:
CLIENT_ID = 'CLN1QMOSO2WMBTVJXM5CB2TBBBBRWEHS4A3ZPI25C045K3ZJ' # your Foursquare ID
CLIENT_SECRET = 'CNFHFFSJUGT0JZQT5PSQB4ERRJQGRU13IMRICEP31O2X10VU' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

LIMIT = 100
radius = 500
#print('Your credentails:')
#print('CLIENT_ID: ' + CLIENT_ID)
#print('CLIENT_SECRET:' + CLIENT_SECRET)

<a id='item2'></a>

## 2. Explore All Neighborhoods in Toronto

#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [17]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
   
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        print (name,"number of venues:", len(results))
       
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [18]:
toronto_data.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [19]:
# type your answer here

toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )



Rouge, Malvern number of venues: 2
Highland Creek, Rouge Hill, Port Union number of venues: 2
Guildwood, Morningside, West Hill number of venues: 9
Woburn number of venues: 4
Cedarbrae number of venues: 8
Scarborough Village number of venues: 1
East Birchmount Park, Ionview, Kennedy Park number of venues: 4
Clairlea, Golden Mile, Oakridge number of venues: 10
Cliffcrest, Cliffside, Scarborough Village West number of venues: 3
Birch Cliff, Cliffside West number of venues: 4
Dorset Park, Scarborough Town Centre, Wexford Heights number of venues: 6
Maryvale, Wexford number of venues: 7
Agincourt number of venues: 4
Clarks Corners, Sullivan, Tam O'Shanter number of venues: 10
Agincourt North, L'Amoreaux East, Milliken, Steeles East number of venues: 3
L'Amoreaux West number of venues: 12
Upper Rouge number of venues: 0
Hillcrest Village number of venues: 4
Fairview, Henry Farm, Oriole number of venues: 66
Bayview Village number of venues: 4
Silver Hills, York Mills number of venues: 1
Newt

#### Let's check the size of the resulting dataframe

In [20]:
print(toronto_venues.shape)

(2254, 7)


Let's check how many venues were returned for each neighborhood

In [21]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      3   
Albion Gardens, Beaumond Heights, Humbergate, J...                     11   
Alderwood, Long Branch                                                 10   
Bathurst Manor, Downsview North, Wilson Heights                        18   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      25   
Berczy Park                                                            55   
Birch Cliff, Cliffside West                                             4   
Bloordale Gardens, Eringate, Markland Wood, Old...                      6   
Brockton, Exhibition Place, Parkdale Village                           21   
Business Reply Mail Processing Centre 969 Eastern                      19   
CFB Toronto, Downsview East                                             3   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            46   
Caledonia-Fairbanks                                                     6   
Canada Post Gateway Processing Centre                                  10   
Cedarbrae                                                               8   
Central Bay Street                                                     88   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               15   
Church and Wellesley                                                   87   
Clairlea, Golden Mile, Oakridge                                        10   
Clarks Corners, Sullivan, Tam O'Shanter                                10   
Cliffcrest, Cliffside, Scarborough Village West                         3   
Cloverdale, Islington, Martin Grove, Princess G...                      1   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             38   
Davisville North                                                        9   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     14   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                          4   
Design Exchange, Toronto Dominion Centre                              100   
Don Mills North                                                         5   
Dorset Park, Scarborough Town Centre, Wexford H...                      6   
Dovercourt Village, Dufferin                                           20   
Downsview Central                                                       4   
Downsview Northwest                                                     5   
Downsview West                                                          4   
Downsview, North Park, Upwood Park                                      5   
East Birchmount Park, Ionview, Kennedy Park                             4   
East Toronto                                                            3   
Emery, Humberlea                                                        1   
Fairview, Henry Farm, Oriole                                           66   
First Canadian Place, Underground city                                100   
Flemingdon Park, Don Mills South                                       23   
Forest Hill North, Forest Hill West                                     4   
Glencairn                                                               4   
Guildwood, Morningside, West Hill                                       9   
Harbord, University of Toronto                                         34   
Harbourfront East, Toronto Islands, Union Statio

#### Let's find out how many unique categories can be curated from all the returned venues

In [22]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 280 uniques categories.


In [23]:
tdf = toronto_venues.groupby('Neighborhood').count()
tdf.shape

(100, 6)

Note: While there are 103 postal codes,  3 of them had no nearby venus. As a result, the dataframe grouped by 'Neighborhodd' has only 100 unique locations/Postal codes.

## 3. Analyze Each Neighborhood

In [24]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe - Note that "Neighborhood" is a category and there is alreay such a column
# I'm naming the new column NH_Name instead
toronto_onehot['NH_Name'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()


NH_Name  Accessories Store  Adult Boutique  \
0                          Rouge, Malvern                  0               0   
1                          Rouge, Malvern                  0               0   
2  Highland Creek, Rouge Hill, Port Union                  0               0   
3  Highland Creek, Rouge Hill, Port Union                  0               0   
4       Guildwood, Morningside, West Hill                  0               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    1               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Burger Joint 

And let's examine the new dataframe size.

In [25]:
toronto_onehot.shape

(2254, 281)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [26]:
toronto_grouped = toronto_onehot.groupby('NH_Name').mean().reset_index()
toronto_grouped

NH_Name  Accessories Store  \
0                            Adelaide, King, Richmond             0.0000   
1                                           Agincourt             0.0000   
2   Agincourt North, L'Amoreaux East, Milliken, St...             0.0000   
3   Albion Gardens, Beaumond Heights, Humbergate, ...             0.0000   
4                              Alderwood, Long Branch             0.0000   
5     Bathurst Manor, Downsview North, Wilson Heights             0.0000   
6                                     Bayview Village             0.0000   
7                   Bedford Park, Lawrence Manor East             0.0000   
8                                         Berczy Park             0.0000   
9                         Birch Cliff, Cliffside West             0.0000   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...             0.0000   
11       Brockton, Exhibition Place, Parkdale Village             0.0000   
12  Business Reply Mail Processing Centre 969 Eastern             0.0000   
13                        CFB Toronto, Downsview East             0.0000   
14  CN Tower, Bathurst Quay, Island airport, Harbo...             0.0000   
15                        Cabbagetown, St. James Town             0.0000   
16                                Caledonia-Fairbanks             0.0000   
17              Canada Post Gateway Processing Centre             0.0000   
18                                          Cedarbrae             0.0000   
19                                 Central Bay Street             0.0000   
20          Chinatown, Grange Park, Kensington Market             0.0000   
21                                           Christie             0.0000   
22                               Church and Wellesley             0.0000   
23                    Clairlea, Golden Mile, Oakridge             0.0000   
24            Clarks Corners, Sullivan, Tam O'Shanter             0.0000   
25    Cliffcrest, Cliffside, Scarborough Village West             0.0000   
26  Cloverdale, Islington, Martin Grove, Princess ...             0.0000   
27                     Commerce Court, Victoria Hotel             0.0000   
28                                         Davisville             0.0000   
29                                   Davisville North             0.0000   
30  Deer Park, Forest Hill SE, Rathnelly, South Hi...             0.0000   
31     Del Ray, Keelesdale, Mount Dennis, Silverthorn             0.0000   
32           Design Exchange, Toronto Dominion Centre             0.0000   
33                                    Don Mills North             0.0000   
34  Dorset Park, Scarborough Town Centre, Wexford ...             0.0000   
35                       Dovercourt Village, Dufferin             0.0000   
36                                  Downsview Central             0.0000   
37                                Downsview Northwest             0.0000   
38                                     Downsview West             0.0000   
39                 Downsview, North Park, Upwood Park             0.0000   
40        East Birchmount Park, Ionview, Kennedy Park             0.0000   
41                                       East Toronto             0.0000   
42                                   Emery, Humberlea             0.0000   
43                       Fairview, Henry Farm, Oriole             0.0000   
44             First Canadian Place, Underground city             0.0000   
45                   Flemingdon Park, Don Mills South             0.0000   
46                Forest Hill North, Forest Hill West             0.0000   
47                                          Glencairn             0.0000   
48                  Guildwood, Morningside, West Hill             0.0000   
49                     Harbord, University of Toronto             0.0000   
50  Harbourfront East, Toronto Islands, Union Station             0.0000   
51                          Harbourfront, Regent Park             0.0000   
52                

#### Let's confirm the new size

In [27]:
toronto_grouped.shape

(100, 281)

#### Let's print each neighborhood along with the top 5 most common venues

In [28]:
num_top_venues = 5

for hood in toronto_grouped['NH_Name']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['NH_Name'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0                 Café  0.05
1          Coffee Shop  0.05
2           Steakhouse  0.04
3                  Bar  0.04
4  American Restaurant  0.04


----Agincourt----
                venue  freq
0      Sandwich Place  0.25
1      Breakfast Spot  0.25
2  Chinese Restaurant  0.25
3              Lounge  0.25
4   Accessories Store  0.00


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
               venue  freq
0         Playground  0.33
1   Asian Restaurant  0.33
2               Park  0.33
3  Accessories Store  0.00
4  Mobile Phone Shop  0.00


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                 venue  freq
0        Grocery Store  0.18
1          Video Store  0.09
2       Sandwich Place  0.09
3  Fried Chicken Joint  0.09
4         Liquor Store  0.09


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place   0.2
1

               venue  freq
0         Playground  0.25
1   Department Store  0.25
2        Coffee Shop  0.25
3     Discount Store  0.25
4  Accessories Store  0.00


----East Toronto----
               venue  freq
0  Convenience Store  0.33
1               Park  0.33
2        Coffee Shop  0.33
3  Accessories Store  0.00
4  Mobile Phone Shop  0.00


----Emery, Humberlea----
               venue  freq
0     Baseball Field   1.0
1  Accessories Store   0.0
2  Mobile Phone Shop   0.0
3      Movie Theater   0.0
4              Motel   0.0


----Fairview, Henry Farm, Oriole----
                  venue  freq
0        Clothing Store  0.12
1  Fast Food Restaurant  0.09
2           Coffee Shop  0.08
3            Restaurant  0.05
4      Toy / Game Store  0.03


----First Canadian Place, Underground city----
                 venue  freq
0          Coffee Shop  0.10
1                 Café  0.08
2                Hotel  0.06
3           Restaurant  0.04
4  American Restaurant  0.03


----Flemingdon Park,

4               Motel   0.0


----Silver Hills, York Mills----
               venue  freq
0          Cafeteria   1.0
1  Accessories Store   0.0
2  Mobile Phone Shop   0.0
3      Movie Theater   0.0
4              Motel   0.0


----St. James Town----
         venue  freq
0         Café  0.06
1  Coffee Shop  0.06
2        Hotel  0.05
3   Restaurant  0.05
4       Bakery  0.03


----Stn A PO Boxes 25 The Esplanade----
                venue  freq
0         Coffee Shop  0.11
1          Restaurant  0.04
2                Café  0.04
3               Hotel  0.03
4  Seafood Restaurant  0.03


----Studio District----
                 venue  freq
0                 Café  0.11
1          Coffee Shop  0.08
2               Bakery  0.05
3            Gastropub  0.05
4  American Restaurant  0.05


----The Annex, North Midtown, Yorkville----
            venue  freq
0            Café  0.13
1     Coffee Shop  0.13
2  Sandwich Place  0.13
3     Pizza Place  0.09
4             Pub  0.04


----The Beaches----
  

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [30]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['NH_Name']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2  Agincourt North, L'Amoreaux East, Milliken, St...                  Park   
3  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café   American Restaurant                   Bar   
1        Breakfast Spot        Sandwich Place    Chinese Restaurant   
2            Playground      Asian Restaurant           Yoga Studio   
3  Fast Food Restaurant           Pizza Place        Sandwich Place   
4           Coffee Shop                   Pub        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Steakhouse        Cosmetics Shop                 Hotel   
1           Yoga Studio   Dumpling Restaurant               Dog Run   
2   Dumpling Restaurant        Discount Store               Dog Run   
3            Beer Store           Coffee Shop          Liquor Store   
4                  Pool          Skating Rink          Dance Studio   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant          Burger Joint                 Bakery  
1      Doner Restaurant            Donut Shop              Drugstore  
2      Doner Restaurant            Donut Shop              Drugstore  
3   Fried Chicken Joint           Video Store               Pharmacy  
4              Pharmacy                   Gym    Dumpling Restaurant

<a id='item4'></a>

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [31]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('NH_Name', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:100] 

array([0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 4, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 4, 3, 0,
       0, 0, 4, 4, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 4, 0, 0, 0, 4, 0, 0,
       0, 2, 0, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 2, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 4, 0, 0, 0, 4, 0, 0, 0, 0, 0, 4])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [32]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster_Labels', kmeans.labels_)
#neighborhoods_venues_sorted

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')


toronto_merged.head() # check the last columns!

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster_Labels       1st Most Common Venue  \
0 -79.194353             0.0        Fast Food Restaurant   
1 -79.160497             0.0  Construction & Landscaping   
2 -79.188711             0.0                Intersection   
3 -79.216917             0.0                 Coffee Shop   
4 -79.239476             0.0            Hakka Restaurant   

  2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0            Print Shop          Dessert Shop                        Diner   
1                   Bar           Yoga Studio  Eastern European Restaurant   
2                   Spa     Electronics Store                  Pizza Place   
3     Convenience Store     Korean Restaurant                  Yoga Studio   
4                Bakery                Lounge         Caribbean Restaurant   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0               Discount Store               Dog Run      Doner Restaurant   
1                      Dog Run      Doner Restaurant            Donut Shop   
2               Breakfast Spot          Tech Startup        Medical Center   
3  Eastern European Restaurant               Dog Run      Doner Restaurant   
4           Athletics & Sports                  Bank   Fried Chicken Joint   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop             Drugstore    Dumpling Restaurant  
1             Drugstore   Dumpling Restaurant      Electronics Store  
2   Rental Car Location    Mexican Restaurant                Dog Run  
3            Donut Shop             Drugstore    Dumpling Restaurant  
4       Thai Restaurant           Yoga Studio              Drugstore

Notes:
Remember there are 3 Neighorhoods that have no nearby venues? they are not included in the clustering and their label has 
no a NaN value. As a result, the join made the reset of the values as float instead of int.

Below I put 5 as a new class (so we have 6 clusters!)  and convert the label value to integer. Integer is required since we use the cluster labels to select map colors

In [33]:
toronto_merged['Cluster_Labels'].fillna(5, inplace=True)
toronto_merged['Cluster_Labels'] = toronto_merged['Cluster_Labels'].astype('int32')
#toronto_merged.info()

Finally, let's visualize the resulting clusters

In [34]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters+1)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

Dense residential areas with lots of restaurants/eateries, pub, grocery stores, markets, gym, malls, pharmacies, banks, medical center...

In [35]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster_Labels  \
0                                       Rouge, Malvern               0   
1               Highland Creek, Rouge Hill, Port Union               0   
2                    Guildwood, Morningside, West Hill               0   
3                                               Woburn               0   
4                                            Cedarbrae               0   
6          East Birchmount Park, Ionview, Kennedy Park               0   
7                      Clairlea, Golden Mile, Oakridge               0   
8      Cliffcrest, Cliffside, Scarborough Village West               0   
9                          Birch Cliff, Cliffside West               0   
10   Dorset Park, Scarborough Town Centre, Wexford ...               0   
11                                   Maryvale, Wexford               0   
12                                           Agincourt               0   
13             Clarks Corners, Sullivan, Tam O'Shanter               0   
15                                     L'Amoreaux West               0   
17                                   Hillcrest Village               0   
18                        Fairview, Henry Farm, Oriole               0   
19                                     Bayview Village               0   
22                                    Willowdale South               0   
24                                     Willowdale West               0   
26                                     Don Mills North               0   
27                    Flemingdon Park, Don Mills South               0   
28     Bathurst Manor, Downsview North, Wilson Heights               0   
29                     Northwood Park, York University               0   
31                                      Downsview West               0   
32                                   Downsview Central               0   
33                                 Downsview Northwest               0   
34                                    Victoria Village               0   
35                     Woodbine Gardens, Parkview Hill               0   
36                                    Woodbine Heights               0   
37                                         The Beaches               0   
38                                             Leaside               0   
39                                    Thorncliffe Park               0   
41                        The Danforth West, Riverdale               0   
42                      The Beaches West, India Bazaar               0   
43                                     Studio District               0   
45                                    Davisville North               0   
46                                  North Toronto West               0   
47                                          Davisville               0   
49   Deer Park, Forest Hill SE, Rathnelly, South Hi...               0   
51                         Cabbagetown, St. James Town               0   
52                                Church and Wellesley               0   
53                           Harbourfront, Regent Park               0   
54                            Ryerson, Garden District               0   
55                                      St. James Town               0   
56                                         Berczy Park               0   
57                                  Central Bay Street               0   
58                            Adelaide, King, Richmond               0   
59   Harbourfront East, Toronto Islands, Union Station               0   
60            Design Exchange, Toronto Dominion Centre               0   
61                      Commerce Court, Victoria Hotel               0   
62                   Bedford Park, Lawrence Manor East               0   
63                                            Roselawn               0   
65                 The Annex, North Midtown, Yorkville               0   
66                      Harbord, University of

#### Cluster 2
Unique from other neighorhoods - simple, quiet residential area?

In [36]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster_Labels 1st Most Common Venue  \
20  Silver Hills, York Mills               1             Cafeteria   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20  Eastern European Restaurant        Discount Store               Dog Run   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20      Doner Restaurant            Donut Shop             Drugstore   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
20   Dumpling Restaurant           Yoga Studio                  Diner

#### Cluster 3
Playground, Dog Run, Yoga Studio, Tenis Court, Donuts, Doner Restaurants, ...

In [37]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster_Labels  \
5                                 Scarborough Village               2   
14  Agincourt North, L'Amoreaux East, Milliken, St...               2   
48                        Moore Park, Summerhill East               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
5             Playground           Yoga Studio   Dumpling Restaurant   
14                  Park            Playground      Asian Restaurant   
48          Tennis Court            Playground           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
5                  Diner        Discount Store               Dog Run   
14           Yoga Studio   Dumpling Restaurant        Discount Store   
48   Dumpling Restaurant                 Diner        Discount Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
5       Doner Restaurant            Donut Shop             Drugstore   
14               Dog Run      Doner Restaurant            Donut Shop   
48               Dog Run      Doner Restaurant            Donut Shop   

         10th Most Common Venue  
5   Eastern European Restaurant  
14                    Drugstore  
48                    Drugstore

#### Cluster 4
Almost identical neighborhood with Baseball fields, Yoga studio, Dog Run, Donor Restaurant, ...


In [38]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster_Labels  \
91  Humber Bay, King's Mill Park, Kingsway Park So...               3   
97                                   Emery, Humberlea               3   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
91        Baseball Field           Yoga Studio               Dog Run   
97        Baseball Field           Yoga Studio               Dog Run   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
91      Doner Restaurant            Donut Shop             Drugstore   
97      Doner Restaurant            Donut Shop             Drugstore   

   7th Most Common Venue        8th Most Common Venue 9th Most Common Venue  \
91   Dumpling Restaurant  Eastern European Restaurant     Electronics Store   
97   Dumpling Restaurant  Eastern European Restaurant     Electronics Store   

   10th Most Common Venue  
91                  Diner  
97                  Diner

#### Cluster 5
Park, Trail, Playgrounds,...

In [39]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster_Labels  \
23                                     York Mills West               4   
25                                           Parkwoods               4   
30                         CFB Toronto, Downsview East               4   
40                                        East Toronto               4   
44                                       Lawrence Park               4   
50                                            Rosedale               4   
64                 Forest Hill North, Forest Hill West               4   
72                                           Glencairn               4   
74                                 Caledonia-Fairbanks               4   
79                  Downsview, North Park, Upwood Park               4   
90       The Kingsway, Montgomery Road, Old Mill North               4   
94   Cloverdale, Islington, Martin Grove, Princess ...               4   
98                                              Weston               4   
100  Kingsview Village, Martin Grove Gardens, Richv...               4   

    1st Most Common Venue 2nd Most Common Venue        3rd Most Common Venue  \
23                   Park     Convenience Store                         Bank   
25   Fast Food Restaurant                  Park            Food & Drink Shop   
30                   Park               Airport            Other Repair Shop   
40                   Park           Coffee Shop            Convenience Store   
44                   Park              Bus Line                  Swim School   
50                   Park            Playground                        Trail   
64                  Trail                  Park             Sushi Restaurant   
72                   Park           Pizza Place          Japanese Restaurant   
74                   Park              Pharmacy                       Market   
79                   Park                Bakery   Construction & Landscaping   
90                  River                  Park                         Pool   
94                   Bank           Yoga Studio  Eastern European Restaurant   
98                   Park           Yoga Studio          Dumpling Restaurant   
100           Pizza Place                  Park            Mobile Phone Shop   

    4th Most Common Venue        5th Most Common Venue 6th Most Common Venue  \
23                    Bar                  Yoga Studio   Dumpling Restaurant   
25    Dumpling Restaurant                        Diner        Discount Store   
30            Yoga Studio  Eastern European Restaurant        Discount Store   
40            Yoga Studio  Eastern European Restaurant        Discount Store   
44            Yoga Studio          Dumpling Restaurant               Dog Run   
50            Yoga Studio          Dumpling Restaurant                 Diner   
64          Jewelry Store                  Yoga Studio        Discount Store   
72                    Pub                  Yoga Studio            Donut Shop   
74   Fast Food Restaurant                Women's Store    Falafel Restaurant   
79          Deli / Bodega             Basketball Court           Yoga Studio   
90             Donut Shop                 Dessert Shop    Dim Sum Restaurant   
94                Dog Run             Doner Restaurant            Donut Shop   
98                  Diner               Discount Store               Dog Run   
100              Bus Line                  Yoga Studio             Drugstore   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
23                       Dog Run      Doner Restaurant            Donut Shop   
25                       Dog Run      Doner Restaurant            Donut Shop   
30                       Dog Run      Doner Restaurant            Donut Shop   
40                       Dog Run      Doner Restaurant            Donut Shop   
44              Doner Restaurant            Donut Shop             Drugstore   
50                Discount Store               Do

#### Cluster 6
Don't forget the three neighborhoods that have no nearby venues!

In [40]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 5, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster_Labels 1st Most Common Venue  \
16              Upper Rouge               5                   NaN   
21  Newtonbrook, Willowdale               5                   NaN   
93         Islington Avenue               5                   NaN   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
16                   NaN                   NaN                   NaN   
21                   NaN                   NaN                   NaN   
93                   NaN                   NaN                   NaN   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
16                   NaN                   NaN                   NaN   
21                   NaN                   NaN                   NaN   
93                   NaN                   NaN                   NaN   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
16                   NaN                   NaN                    NaN  
21                   NaN                   NaN                    NaN  
93                   NaN                   NaN                    NaN

#### This marks the end of the exercise